File: 701/python/solB.py
import sys
import argparse


def main():
    input = sys.stdin.readline
    n,q,k = list(map(int,input().split()))
    a = list(map(int,input().split()))

    # preprocessing
    pre = [0]*n
    for i in range(n):
        if i >= 1:
            pre[i] = pre[i-1]
        if i == 0:
            pre[i] += a[i+1]-2
        elif i == n-1: 
            pre[i] += a[i-1]-2
        else:   
            pre[i] += a[i+1]-a[i-1]-2
    for i in range(q):
        l,r = list(map(int,input().split()))
        l,r = l-1, r-1
        ans = 0
        if l == r:
            ans = k-1
        else:
            ans += pre[r-1]-pre[l]
            # x = pre[r-1]-pre[l+1] if l+1 and r-1 is in a good range
            ans += a[l+1]-2
            ans += k-a[r-1]-1
        print(ans)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('--file', dest='filename', default=None)
    args = parser.parse_args()
    if args.filename is not None:
        sys.stdin = open(args.filename)
    main()

File: 2020-2021 ICPC, NERC, Northern Eurasia Onsite/python/solb.cpp
import sys

def op1(n, x):
    # 0 -> 1, 2 -> 3
    for i in range(0,n):
        u = 2*i
        v = 2*i+1
        t1 = x[u]
        x[u] = x[v]
        x[v] = t1
    return x

def op2(n,x):
    # 0 -> n 1 -> n+1
    for i in range(0,n):
        u = i
        v = i+n
        t1 = x[u]
        x[u] = x[v]
        x[v] = t1
    return x

def main():
    n, q  =list(map(int,input().split()))
    for _ in range(q):
        m = 2*n
        x = [i for i in range(m)]
        b = list(map(int,input().split()))
        for i in range(len(b)):
            if b[i] == 0:
                x = op1(n,x)
            if b[i] == 1:
                x = op2(n,x)
        print(x)

    pass

if __name__ == "__main__":
    input = sys.stdin.readline
    main()